src/AppInstallerRepositoryCore/CompositeSource.cpp

```diff
@@ -247,6 +248,233 @@ namespace AppInstaller::Repository
             return installedVersion;
         }
 
+        HRESULT CheckInstalledLocationStatus(const std::filesystem::path& installedLocation)
+        {
+            HRESULT installLocationStatus = WINGET_INSTALLED_STATUS_INSTALL_LOCATION_NOT_APPLICABLE;
+            if (!installedLocation.empty())
+            {
+                // Use the none throw version, if the directory cannot be reached, it's treated as not found and later file checks are not performed.
+                std::error_code error;
+                installLocationStatus =
+                    std::filesystem::exists(installedLocation, error) && std::filesystem::is_directory(installedLocation, error) ?
+                    WINGET_INSTALLED_STATUS_INSTALL_LOCATION_FOUND :
+                    WINGET_INSTALLED_STATUS_INSTALL_LOCATION_NOT_FOUND;
+            }
+
+            return installLocationStatus;
+        }
+
+        // Map to cache already calculated file hashes.
+        struct FilePathComparator
+        {
+            bool operator()(const std::filesystem::path& a, const std::filesystem::path& b) const
+            {
+                if (std::filesystem::equivalent(a, b))
+                {
+                    return false;
+                }
+
+                return a < b;
+            }
+        };
+        using FileHashMap = std::map<std::filesystem::path, Utility::SHA256::HashBuffer, FilePathComparator>;
+
+        HRESULT CheckInstalledFileStatus(
+            const std::filesystem::path& filePath,
+            const Utility::SHA256::HashBuffer& expectedHash,
+            FileHashMap& fileHashes)
+        {
+            HRESULT fileStatus = WINGET_INSTALLED_STATUS_FILE_NOT_FOUND;
+            try
+            {
+                if (std::filesystem::exists(filePath) && std::filesystem::is_regular_file(filePath))
+                {
+                    fileStatus = WINGET_INSTALLED_STATUS_FILE_FOUND_WITHOUT_HASH_CHECK;
+                    if (!expectedHash.empty())
+                    {
+                        if (fileHashes.find(filePath) == fileHashes.end())

```

Comment:
If we are going to have a map for performance, then you should:
```
auto itr = find();
if (itr != end)
{
itr = emplace().first;
}

AreEqual(expected, itr->second);
```
To prevent any more map lookups than necessary.